akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1342,3 +1342,47 @@ abstract class AbstractOutHandler extends OutHandler
  * (completing when upstream completes, failing when upstream fails, completing when downstream cancels).
  */
 abstract class AbstractInOutHandler extends InHandler with OutHandler
+
+/**
+ * This trait wraps callback for `GraphStage` stage instances and handle gracefully cases when stage is
+ * not yet initialized or already finished.
+ *
+ * While `GraphStage` has not initialized it adds all requests to list.
+ * As soon as `GraphStage` is started it stops collecting requests (pointing to real callback
+ * function) and run all the callbacks from the list
+ *
+ * Supposed to be used by GraphStages that share call back to outer world
+ */
+private[akka] trait CallbackWrapper[T] {
+  private trait CallbackState
+  private case class NotInitialized(list: List[T]) extends CallbackState
+  private case class Initialized(f: T ⇒ Unit) extends CallbackState
+  private case class Stopped(f: T ⇒ Unit) extends CallbackState
+
+  private val callbackState = new AtomicReference[CallbackState](NotInitialized(Nil))
+
+  def stopCallback(f: T ⇒ Unit): Unit = {
+    callbackState.set(Stopped(f))
+  }
+
+  def initCallback(f: T ⇒ Unit): Unit = {
+    val list = (callbackState.getAndSet(Initialized(f)): @unchecked) match {
+      case NotInitialized(l) ⇒ l
+    }
+    list.reverse.foreach(f)
+  }
+
+  def callback(arg: T): Unit = {
+    callbackState.get() match {
+      case Initialized(cb) ⇒ cb(arg)
+      case list @ NotInitialized(l) ⇒
+        if (!callbackState.compareAndSet(list, NotInitialized(arg :: l)))
+          callbackState.get() match {
+            case NotInitialized(_) ⇒ throw new IllegalStateException("Concurrent call of CallbackWrapper is detected")
```

Comment:
This is too limiting for the general case, but it is desirable for QueueSink. I think I’d rather remove it because concurrent use goes unnoticed in almost all cases anyway.